### PR TITLE
Added encoding/decoding

### PR DESCRIPTION
**Pull request by umbreak**

### CODE REVIEW

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/acls/AclState.scala [hunk 1]
```diff
@@ -33,8 +33,8 @@ object AclState {
                            rev: Long,
                            created: Instant,
```
Comment (bogdanromanx) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/acls/AclState.scala]: we should update these to: `createdAt` and `updatedAt` to remain consistent with the `ResourceF[A]`

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/types/Caller.scala [hunk 2]
```diff
@@ -0,0 +1,4 @@
+package ch.epfl.bluebrain.nexus.iam.types
+import ch.epfl.bluebrain.nexus.iam.types.Identity.Subject
+
+final case class Caller(subject: Subject, identities: Set[Identity])
```
Comment (bogdanromanx) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/types/Caller.scala]: docs? it's also important to note if the subject is also included in the identity set.

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/types/Permission.scala [hunk 3]
```diff
@@ -14,7 +14,10 @@ object Permission {
   private val valid = "[a-zA-Z-:_\\/]{1,32}".r
 
   final def apply(value: String): Option[Permission] =
-    valid.findFirstIn(value).map(new Permission(_))
+    valid.findFirstIn(value).map(unsafe)
+
+  final def unsafe(value: String): Permission =
```
Comment (bogdanromanx) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/types/Permission.scala]: docs

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/types/ResourceF.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    rev: Long,
+    types: Set[AbsoluteIri],
+    createdAt: Instant,
+    createdBy: Identity,
+    updatedAt: Instant,
+    updatedBy: Identity,
+    value: A
+) {
+
+  def map[B](f: A => B): ResourceF[B] =
```
Comment (bogdanromanx) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/types/ResourceF.scala]: docs

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/types/ResourceF.scala [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    copy(value = f(value))
+
+  def discard: ResourceF[Unit] =
+    map(_ => ())
+}
+
+object ResourceF {
+  def unit(
```
Comment (bogdanromanx) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/types/ResourceF.scala]: docs

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/types/ResourceMetadata.scala [hunk 6]
```diff
@@ -2,24 +2,19 @@ package ch.epfl.bluebrain.nexus.iam.types
...
-final case class ResourceMetadata(id: AbsoluteIri,
-                                  rev: Long,
-                                  types: Set[AbsoluteIri],
-                                  createdBy: Identity,
-                                  updatedBy: Identity,
-                                  createdAt: Instant,
-                                  updatedAt: Instant)
+object ResourceMetadata {
+
+  def apply(
```
Comment (bogdanromanx) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/types/ResourceMetadata.scala]: docs

**Comment by umbreak** [comment]
I don't want to delete the modules directory yet (let's wait until we finish `Realms`). I just removed the changes on those files (since it wasn't required, just my mistake when doing replace)

**Pull request merged**